dlt/destinations/impl/sink/sink.py

```diff
@@ -0,0 +1,175 @@
+from abc import ABC, abstractmethod
+from types import TracebackType
+from typing import ClassVar, Dict, Optional, Type, Iterable, Iterable
+from dlt.common.configuration import with_config, known_sections
+
+from dlt.destinations.job_impl import EmptyLoadJob
+from dlt.common.typing import TDataItems
+from dlt.common import json
+from dlt.pipeline.current import (
+    destination_state,
+    commit_load_package_state,
+)
+
+from dlt.common.schema import Schema, TTableSchema, TSchemaTables
+from dlt.common.schema.typing import TTableSchema
+from dlt.common.storages import FileStorage
+from dlt.common.destination import DestinationCapabilitiesContext
+from dlt.common.destination.reference import (
+    TLoadJobState,
+    LoadJob,
+    JobClientBase,
+)
+
+from dlt.destinations.impl.sink import capabilities
+from dlt.destinations.impl.sink.configuration import SinkClientConfiguration, TSinkCallable
+
+
+class SinkLoadJob(LoadJob, ABC):
+    def __init__(
+        self,
+        table: TTableSchema,
+        file_path: str,
+        config: SinkClientConfiguration,
+        schema: Schema,
+        destination_state: Dict[str, int],
+        resolved_callable: TSinkCallable,
+    ) -> None:
+        super().__init__(FileStorage.get_file_name_from_file_path(file_path))
+        self._file_path = file_path
+        self._config = config
+        self._table = table
+        self._schema = schema
+        self._resolved_callable = resolved_callable
+
+        self._state: TLoadJobState = "running"
+        self._storage_id = f"{self._parsed_file_name.table_name}.{self._parsed_file_name.file_id}"
+        try:
+            if self._config.batch_size == 0:
+                # on batch size zero we only call the callable with the filename
+                self.call_callable_with_items(self._file_path)
+            else:
+                current_index = destination_state.get(self._storage_id, 0)
+                for batch in self.run(current_index):
+                    self.call_callable_with_items(batch)
+                    current_index += len(batch)
+                    destination_state[self._storage_id] = current_index
```

Comment:
are you able to store batch number per table in state? (you'll need to lock the state). then user will be able to figure out that a batch is a first batch for a given table. you already store stuff in state